### PR TITLE
Various tweaks to command argument handling

### PR DESCRIPTION
**Pull request by zml2008**
[replacing #784]

This PR covers a few long-standing issues about command argument format and hopefully command argument handling, as well as refactoring some annoying points of using CommandSpec and CommandElements.

The ones I want to get are:
- [x] #604 
- [x] #583 
- [x] Adding more tags to the various location-related command arguments
- [x] #689 -- An event that is called for chat as well as commands (very preprocessey) -- completed, this is covered by #957
- [x] #780 
- [x] ~~Only parse escapes for characters that would otherwise have a special function~~
- [ ] Refactor CommandArgs + CommandContext to be less dumb (see [PEX wip on this](https://github.com/PEXPlugins/PermissionsEx/tree/feature/elementresult) (though what happens from this will be different))
- [ ] #924 
- [x] #989 

notable todos:
- Implement `#spawn:<world>` on LocationCommandElement
- Properly handle tabcompletion on LocationCommandElement

### CODE REVIEW

File: src/main/java/org/spongepowered/api/command/args/CommandElement.java [hunk 1]
```diff
@@ -131,7 +71,81 @@ public void parse(CommandSource source, CommandArgs args, CommandContext context
      * @param src The source requesting usage
      * @return The formatted usage
      */
-    public Text getUsage(CommandSource src) {
-        return getKey() == null ? Text.of() : Text.of("<", getKey(), ">");
+    public abstract Text getUsage(CommandSource src);
+
+    public static abstract class Value<T> extends CommandElement {
+        private final Text key;
```
Comment (CrystalMageX) [inline src/main/java/org/spongepowered/api/command/args/CommandElement.java:77]: This is either missing a `Nullable` annotation, or the constructor is missing a null check.

**Comment by VcSaJen** [comment]
I have no idea if it was already fixed in one of PRs, but...

When command with subcommands is inputted by user incorrectly, currently it displays usage of main command (I.e. only list of subcommands).

For example, command "mycmd" have subcommand "change", which in turn have subcommand "coords". User inputs "/mycmd change argblarg":

It would be more useful if error would display list of valid child commands of "change" (coords, etc), not a list of child commands of the root command.

User inputs "/mycmd change coords too many a r g u m e n t s":

Error should display usage of the end command "/mycmd change coords" (I.e. it's parameters, etc).

**Comment by phroa** [comment]
While you're changing things, I'm not sure I like the `cc` style error indicator.

Given the (unfortunate) variable-spaced font, I think it would be better to highlight the broken section with a pink? colour. It's really ambiguous where the `^` points to sometimes.

File: src/main/java/org/spongepowered/api/command/args/ArgUtils.java [hunk 2]
```diff
@@ -22,43 +22,31 @@
...
 
-    /**
-     * Returns an input tokenizer that returns the input string as a single argument.
-     *
-     * @return The appropriate tokenizer
-     */
-    public static InputTokenizer rawInput() {
-        return RawStringInputTokenizer.INSTANCE;
+    @Nullable
+    public static String textToArgKey(@Nullable Text key) {
```
Comment (phroa) [inline src/main/java/org/spongepowered/api/command/args/ArgUtils.java:40]: If these are internal, maybe this shouldn't be public? /pedantic

File: src/main/java/org/spongepowered/api/command/args/ChildCommand.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    }
+
+    public static Map.Entry<CommandElement.Value<Optional<ChildCommand>>, CommandExecutor> forChildrenWithFallback(Map<List<String>, CommandSpec>
+            children, CommandExecutor fallbackExecutor) {
+        checkNotNull(fallbackExecutor, "fallbackExecutor");
+
+        CommandElement.Value<Optional<ChildCommand>> element = optional(createElement(children));
+        CommandExecutor ex =  (src, args) -> {
+            Optional<ChildCommand> val = args.get(element);
```
Comment (phroa) [inline src/main/java/org/spongepowered/api/command/args/ChildCommand.java:77]: Could fold this into a `.orElse` to shave a few lines

**Comment by phroa** [comment]
A lot of your lambdas have two spaces before the param list.

**Comment by Zidane** [comment]
@dualspiral 

Closing this as this should be handled, if necessary, in the commands PR.

**Pull request closed without merging**